jupyter_rfb/widget.py

```diff
@@ -257,26 +264,60 @@ def _rfb_maybe_draw(self):
                 if array is not None:
                     self._rfb_send_frame(array)
 
-    def _rfb_send_frame(self, array):
+    def _rfb_schedule_lossless_draw(self, array, delay=0.3):
+        self._rfb_cancel_lossless_draw()
+        loop = asyncio.get_event_loop()
+        handle = loop.call_later(delay, self._rfb_lossless_draw)
+        self._rfb_lossless_draw_info = array, handle
+
+    def _rfb_cancel_lossless_draw(self):
+        if self._rfb_lossless_draw_info:
+            _, handle = self._rfb_lossless_draw_info
+            self._rfb_lossless_draw_info = None
+            handle.cancel()
+
+    def _rfb_lossless_draw(self):
+        array, _ = self._rfb_lossless_draw_info
+        self._rfb_send_frame(array, True)
```

Comment:
This is most of the logic for "ending with a PNG". When a frame is encoded with JPEG, a lossless redraw is scheduled. And this scheduled draw is cancelled in a few places when it's no longer needed.